src/io.ts

```diff
@@ -227,30 +179,22 @@ export function getTodoBatches(
 /**
  * Applies todo changes, either adding or removing, based on batches from `getTodoBatches`.
  *
- * @param todoStorageDir - The .lint-todo storage directory.
+ * @param todoStorageFilePath - The .lint-todo storage file path.
  * @param add - Batch of todos to add.
  * @param remove - Batch of todos to remove.
+ * @param options - An object containing write options.
  */
 export function applyTodoChanges(
-  todoStorageDir: string,
-  add: Map<TodoFileHash, TodoDataV2>,
-  remove: Map<TodoFileHash, TodoDataV2>
+  todoStorageFilePath: string,
+  add: Set<TodoDataV2>,
+  remove: Set<TodoDataV2>,
+  options: Partial<WriteTodoOptions>
 ): void {
-  for (const [fileHash, todoDatum] of add) {
-    const { dir } = posix.parse(fileHash);
-
-    ensureDirSync(posix.join(todoStorageDir, dir));
-    writeJsonSync(posix.join(todoStorageDir, `${fileHash}.json`), todoDatum);
-  }
+  const ops = buildTodoOperations(add, remove);
 
-  for (const [fileHash] of remove) {
-    const { dir } = posix.parse(fileHash);
-    const todoDir = posix.join(todoStorageDir, dir);
-
-    unlinkSync(posix.join(todoStorageDir, `${fileHash}.json`));
-
-    if (readdirSync(todoDir).length === 0) {
-      rmdirSync(todoDir);
-    }
+  if (options.overwrite) {
+    writeFileSync(todoStorageFilePath, ops);
```

Comment:
This is effectively compaction right?

Reply:
Yep. I did implement a separate function to compact though, so I may remove this.